src/escape.rs

```diff
@@ -266,7 +266,7 @@ where
                     unescaped = Some(String::with_capacity(raw.len()));
                 }
                 let unescaped = unescaped.as_mut().expect("initialized");
-                unescaped.push_str(&raw[last_end..start]);
+                unescaped.push_str(&normalize_line_end(&raw[last_end..start]));
```

Comment:
I don't know how to do it. What if there is nothing to unescape? You still need to call `normalize_line_end` for such input in the end, then what is the correct function parameter for `normalize_line_end` so that it can handle both situations?

Reply:
Hm, don't understand what the trouble? `unescape` will always be called when we request data via API methods. If cycling search of `&`, `;`, or `\r` will not find anything, than it is really nothing to change in the data (I assume that cycles for unescaping and normalization merged in one). 

Even if did not merge loops, `normalize_line_end` can accept `&mut Cow<str>` and change it instead of creating an own `Cow` by taking reference to `&str` and returning its own `Cow<str>`

Reply:
I am talking about the not  merge loops, as you point out that there are more line end situations to consider, it is beyond my knowledge to do SIMD stuff myself to merge the loops.

Current function signature is: `fn normalize_line_end(input: &str) -> Cow<str>`
Change to what? How to call it?

Maybe your requirement is beyond my knowledge. In that case, I cannot help. We can close this PR.

Reply:
Ah, read my `&mut Cow<str>` as `&mut Option<String>`. So the whole code will look like:
```rust
fn unescape_with(...) {
  let mut unescaped = None;
  for ... {
    ...
    if need to unescape {
      if unescaped.is_none() {
        unescaped = Some(String::with_capacity(raw.len()));
      }
      let unescaped = unescaped.as_mut().expect("initialized");
    }
    ...
    normalize_line_end(&mut unescaped, raw.len());
    ...
  }
}

fn normalize_line_end(input: &mut Option<String>, raw_len: usize) {
  if need to normalize {
    if input.is_none() {
       input = Some(String::with_capacity(raw_len))
    }
    let normalized = input.as_mut().expect("initialized");
    ...
  }
}
```
> it is beyond my knowledge to do SIMD stuff myself to merge the loops.
Actually, SIMD and merging loops are two separate conception, it is not necessary to explicitly apply SIMD techniques if you not aware of them, it's totally fine. I only tried to be helpful if you aware of it.

> Maybe your requirement is beyond my knowledge. In that case, I cannot help. We can close this PR.

The requirement in only is not to make things worse. First of all we should not make it impossible to correctly process input (that is why we should at least process `\r\u0085` because if we turn it into `\n\u0085`, the user will see two newlines in data instead of one, because specification allows `\n`, `\r\u0085` and `\u0085` be served as (one) newline characters). The second, we should to try that without much impact to performance, so if we can avoid unnecessary allocations we should do that.

If you're having a hard time figuring out how to get there, that's fine, it could be so. Thanks for opening PR anyway. I will consider it later when I return to active phase on the project (currently working on other project) and see how to improve it.

Reply:
Thanks for showing the code.
Yes, it works if you have something to unescape. But what if there is nothing to unescape from the original input?
Then you still have to call normlize_line_end to normalize the original input. How to you put original input raw str as parameter to your current unescape function?

Reply:
> Yes, it works if you have something to unescape. But what if there is nothing to unescape from the original input?
  
Unescape function decides where it is need to unescape something or not. It itself called always (well, because user requests unescaped and normalized data by demand, it will be called by demand, but that "demand" means "unesсape [and normalize] if there is anything")

> How to you put original input raw str as parameter to your current unescape function?

Cannot understand what confuses you here. Calls to `unescape` will not be changed and it is already called whenever necessary.

Reply:
```
pub fn unescape_with<'input, 'entity, F>(
    raw: &'input str,
    mut resolve_entity: F,
) -> Result<Cow<'input, str>, EscapeError>
where
    // the lifetime of the output comes from a capture or is `'static`
    F: FnMut(&str) -> Option<&'entity str>,
{
```

    
Nothing to escape means there is no `&`  and `;` in raw input.  It will not go inside `while` below. So normalize will not happen to input here.
    
```
while let Some(start) = iter.by_ref().find(|p| bytes[*p] == b'&') {
        match iter.next() {
            Some(end) if bytes[end] == b';' => {
                // append valid data
                if unescaped.is_none() {
                    unescaped = Some(String::with_capacity(raw.len()));
                }
                let unescaped = unescaped.as_mut().expect("initialized");
                unescaped.push_str(&normalize_line_end(&raw[last_end..start]));

                // search for character correctness
                let pat = &raw[start + 1..end];
                if let Some(entity) = pat.strip_prefix('#') {
                    let codepoint = parse_number(entity).map_err(EscapeError::InvalidCharRef)?;
                    unescaped.push_str(codepoint.encode_utf8(&mut [0u8; 4]));
                } else if let Some(value) = resolve_entity(pat) {
                    unescaped.push_str(value);
                } else {
                    return Err(EscapeError::UnrecognizedEntity(
                        start + 1..end,
                        pat.to_string(),
                    ));
                }

                last_end = end + 1;
            }
            _ => return Err(EscapeError::UnterminatedEntity(start..raw.len())),
        }
    }
```
Will not go inside here either because nothing has been unescaped

    if let Some(mut unescaped) = unescaped {    
        if let Some(raw) = raw.get(last_end..) {
            unescaped.push_str(&normalize_line_end(raw));
        }
        Ok(Cow::Owned(unescaped))
    } else {

Must run normalize here, so how to call your changed function `fn normalize_line_end(input: &mut Option<String>, raw_len: usize)` here with raw as parameter?
```
        Ok(normalize_line_end(raw))
    }
}
```